### PR TITLE
add container testcases

### PR DESCRIPTION
**Pull request by AhmedHanafy725**
- Test case for creating container with zerotier network
- Test case for adding mount directory to container
- Test case for adding port forward to container
- Test case for adding init process to container
- Test case for creating container with same port forward
- Test case for creating container with host network parameter

### Result:
image

### CODE REVIEW

File: tests/controller/templates_manager/local_temp/container.py [hunk 1]
```diff
@@ -60,4 +60,9 @@ def start(self):
...
     def stop(self):
         return self.service.schedule_action('stop')
-    
\ No newline at end of file
+
+    def add_nic(self, nic):
+        return self.service.schedule_action('add_nic', args={'nic': nic}).wait(die=True)
+
+    def remove_nic(self, name):
+        return self.service.schedule_action('add_nic', args={'name': name})
```
[the remark this answers is not in the document]
Comment (AhmedHanafy725) [reply tests/controller/templates_manager/local_temp/container.py]: done

File: tests/testcases/a_basics/container_testcases.py [hunk 2]
```diff
@@ -49,23 +61,216 @@ def test003_start_stop_container(self):
...
+        container.start()
         time.sleep(2)
 
         self.log('Check if the container has been started, should succeed')
-        conts = self.controller.node.client.container.list()
-        self.assertTrue([c for c in conts.values() if c['container']['arguments']['hostname'] == self.controller.container_manager.data['hostname']])
+        conts = self.node.client.container.list()
+        self.assertTrue([c for c in conts.values() if c['container']['arguments']['hostname'] == container.data['hostname']])
+
+    @parameterized.expand(['before', 'after'])
```
Comment (Dinaamagdy) [inline tests/testcases/a_basics/container_testcases.py]: need more description for testcase options (before and after)

File: tests/testcases/a_basics/container_testcases.py [hunk 3]
```diff
@@ -49,23 +61,216 @@ def test003_start_stop_container(self):
...
+    @unittest.skip('https://github.com/threefoldtech/0-templates/issues/201')
+    def test004_create_container_with_zt_nic(self, add):
+        """ ZRT-ZOS-016
+        *Test case for creating container with zerotier network*
+        **Test Scenario:**
+
+        #. Create a container (C1).
+        #. Add zerotier network to container (c1).
+        #. Check that netwrok is added successfully to the container.
```
Comment (Dinaamagdy) [inline tests/testcases/a_basics/container_testcases.py]: Add check that container has been already joined the zerotier network form zerotier side

File: tests/testcases/a_basics/container_testcases.py [hunk 4]
```diff
@@ -49,23 +61,216 @@ def test003_start_stop_container(self):
...
+        """
+        self.logger.info('Create a container (C1)')
+        zt_network = [{'name': self.random_string(), 'type': 'zerotier', 'id': self.zt_id}, {'type': 'default', 'name': 'defaultnic'}]
+        container = self.controller.container_manager(parent=self.controller, service_name=None)
+        if add == 'before':
+            zt_network[0]['token'] = self.config['zt']['zt_client']
+            container.install(wait=True, nics=zt_network)
```
Comment (Dinaamagdy) [inline tests/testcases/a_basics/container_testcases.py]: add logs to define before or after deployment

File: tests/testcases/a_basics/container_testcases.py [hunk 5]
```diff
@@ -49,23 +61,216 @@ def test003_start_stop_container(self):
...
+        self.log('Create a container (C1) with port forward (P1).')
+        host_port = randint(7000, 8000)
+        guest_port = randint(3000, 4000)
+        container = self.controller.container_manager(parent=self.controller, service_name=None)
+        container.install(wait=True, ports=['{}:{}'.format(host_port, guest_port)])
+        self.assertTrue(container.install_state, " Installtion state is False")
+
+        self.log('Create another container (C2) with same port forward (P1), should fail.')
+        with self.assertRaises(RuntimeError) as e:
+            container1 = self.controller.container_manager(parent=self.controller, service_name=None)
```
Comment (Dinaamagdy) [inline tests/testcases/a_basics/container_testcases.py]: change name to be container 2

File: tests/testcases/a_basics/container_testcases.py [hunk 6]
```diff
@@ -49,23 +61,216 @@ def test003_start_stop_container(self):
...
+            container1.install(wait=True, ports=['{}:{}'.format(host_port, guest_port)])
+
+        self.log('Stop container (C1).')
+        container.stop()
+
+        self.log('Create another container (C3) with same port forward (P1), should success.')
+        container2 = self.controller.container_manager(parent=self.controller, service_name=None)
```
Comment (Dinaamagdy) [inline tests/testcases/a_basics/container_testcases.py]: container3

File: tests/testcases/a_basics/container_testcases.py [hunk 7]
```diff
@@ -49,23 +61,216 @@ def test003_start_stop_container(self):
...
+        self.log('Get container id and client.')
+        conts = self.node.containers.list()
+        cont = [c for c in conts if container.data['hostname'] in c.hostname][0]
+        client = self.node.client.container.client(cont.id)
 
-        
\ No newline at end of file
+        self.log('Try to reach the network, should fail.')
+        result = client.bash('ping 8.8.8.8 -w 5').get()
+        if host_network == 'True':
```
Comment (Dinaamagdy) [inline tests/testcases/a_basics/container_testcases.py]: add logs define that it false or true

**Pull request merged**